pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,48 +70,82 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
+	}
+
+	if err := validation.ValidateNodeResourcesBalancedAllocationArgs(nil, args); err != nil {
+		return nil, err
+	}
+
+	resToWeightMap := make(resourceToWeightMap)
+	for _, resource := range (*args).Resources {
+		resToWeightMap[v1.ResourceName(resource.Name)] = resource.Weight
+	}
+
 	return &BalancedAllocation{
 		handle: h,
 		resourceAllocationScorer: resourceAllocationScorer{
 			BalancedAllocationName,
 			balancedResourceScorer,
-			defaultRequestedRatioResources,
+			resToWeightMap,
 		},
 	}, nil
 }
 
-// todo: use resource weights in the scorer function
 func balancedResourceScorer(requested, allocable resourceToValueMap, includeVolumes bool, requestedVolumes int, allocatableVolumes int) int64 {
-	cpuFraction := fractionOfCapacity(requested[v1.ResourceCPU], allocable[v1.ResourceCPU])
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// This to find a node which has most balanced CPU, memory and volume usage.
-	if cpuFraction >= 1 || memoryFraction >= 1 {
-		// if requested >= capacity, the corresponding host should never be preferred.
-		return 0
+	resourceToFraction := make(map[v1.ResourceName]float64)
+	var totalFraction float64
+	for name, value := range requested {
+		if _, ok := allocable[name]; ok {
+			fraction := fractionOfCapacity(value, allocable[name])
+			if fraction >= 1 {
+				// if requested >= capacity, the corresponding host should never be preferred.
+				return 0
+			}
+			totalFraction = totalFraction + fraction
+			resourceToFraction[name] = fraction
+		} else {
+			klog.InfoS("requested resource not found for node score calculation", "resource", name)
+			return 0
+		}
 	}
 
 	if includeVolumes && utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && allocatableVolumes > 0 {
-		volumeFraction := float64(requestedVolumes) / float64(allocatableVolumes)
+		volumeFraction := fractionOfCapacity(int64(requestedVolumes), int64(allocatableVolumes))
 		if volumeFraction >= 1 {
 			// if requested >= capacity, the corresponding host should never be preferred.
 			return 0
 		}
-		// Compute variance for all the three fractions.
-		mean := (cpuFraction + memoryFraction + volumeFraction) / float64(3)
-		variance := float64((((cpuFraction - mean) * (cpuFraction - mean)) + ((memoryFraction - mean) * (memoryFraction - mean)) + ((volumeFraction - mean) * (volumeFraction - mean))) / float64(3))
-		// Since the variance is between positive fractions, it will be positive fraction. 1-variance lets the
-		// score to be higher for node which has least variance and multiplying it with `MaxNodeScore` provides the scaling
-		// factor needed.
-		return int64((1 - variance) * float64(framework.MaxNodeScore))
+		resourceToFraction["volume"] = volumeFraction
+		totalFraction = totalFraction + volumeFraction
 	}
 
+	// For backward compatibility, if there is only cpu and memory, we will use the legacy formula to calculate the score.
```

Comment:
Why? the formula should be the same

Reply:
Not exactly, the formula we were using for the below the two cases are never the same, 

- cpu and memory:  `(1 - math.Abs(cpuFraction - memoryFraction)) * 100`
- cpu, memory and volume: `(1 - (((cpuFraction -mean) * (cpuFraction -mean) + ...)  / len(resources))) * 100`


let's see a example, suppose `cpuFraction = 0.8, memFraction = 0.5`

It was, 
`diff = 0.8 - 0.5 = 0.3, score = (1 - 0.3) * 100 = 70`

now (same with the case when volume is counted):
` mean = 0.8 + 0.5 / 2 = 0.65
  variance = (0.8 -  0.65) *  (0.8 -  0.65) + (0.65 - 0.5) *  (0.65 - 0.5)  / 2 = 0.0225
  score = (1 - 0.0225) * 100 = 97.75
`

Reply:
So for CPU and memory we just use the difference, whereas for others we use the variance.

Since the variance is not really established (we only use it for the feature gate `BalanceAttachedNodeVolumes`, which is alpha) I'm fine with changing it.

If we were to use the standard deviation, the 2 resources case would go from `1-abs(a-b)` to `1-abs(a-b)/2`. This effectively reduces the effect of balanced allocation. But the difference is better than if we use the variance.

For the example above: `a=0.8`, `b=0.5`
Old formula: `1-0.3=0.7`
Variance:      `1-0.15*0.15`=`0.9775`
STD:             `1-0.15=0.85`

I'm tempted to go STD for consistency, even though it somewhat breaks backwards compatibility. @Huang-Wei ?
Or to preserve behavior we can use: median absolute deviation or range https://en.wikipedia.org/wiki/Statistical_dispersion

Reply:
I put some examples:

![image](https://user-images.githubusercontent.com/1299064/118288888-cc497580-b4a2-11eb-88bd-9812ad1b64bf.png)

The variances don't give a good differentiation among nodes at all. The STD is much better.

Both Range and Mean Average Deviation would preserve the same formula for 2 resources. Between them you can see a clear difference between (1, 1, 0) and (1, 0.5, 0). Do you think is there a reason one of this nodes over the other? If so, we should go with Mean Average Deviation. If not, we should use range.

Reply:
Firstly I think the original formula for 3 resources was implemented incorrectly - it was supposed to return the STD value but forgot to do a final root operation. in @chendave's example, it should return `1 - root of 0.0225`, which is 0.85.

So basically I'm fine with using the STD formula.

> Do you think there is a reason to prefer one of these nodes over the other?

I think the (1, 0.5, 0) is slightly better than (1, 1, 0), and hence MAD is better than RANGE.

So with all the above said, the option we should choose is STD (2-resource case may be incompatible) or MAD.

Reply:
I think I'm fine making the 2 resource case have a smaller difference. This plugin shouldn't have as much influence as other scoring plugins.

Reply:
will switch to STD, thanks for those fantastic inputs!